File: Python3/Partition Equal Subset Sum.py
# 912ms, 40.85%
class Solution:
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        nsum = sum(nums)
        if nsum & 1: return False
        nsum >>= 1
        dp = [True] + [False] * nsum
        for n in nums:
            for i in range(nsum, n - 1, -1):
                dp[i] |= dp[i - n]
        return dp[nsum]
    

File: Cpp/Partition Equal Subset Sum.cc
// 22ms, 81.68%
class Solution {
public:
    bool canPartition(const vector<int> &nums) {
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if (sum & 1) return false;
        sum >>= 1;
        bool dp[sum + 1] = {1};
        for (auto n : nums)
            for (int i = sum; i >= n; --i)
                dp[i] |= dp[i - n];
        return dp[sum];
    }
};
